src/main/java/com/firebase/geofire/GeoQuery.java

```diff
@@ -88,9 +97,10 @@ public synchronized void onCancelled(DatabaseError databaseError) {
     };
 
     private final GeoFire geoFire;
-    private final Set<GeoQueryEventListener> eventListeners = new HashSet<>();
-    private final Map<GeoHashQuery, Query> firebaseQueries = new HashMap<GeoHashQuery, Query>();
+    private final Set<DatasnapshotGeoQueryEventListener> eventListeners = new HashSet<>();
+    private final Map<GeoHashQuery, Query> firebaseQueries = new HashMap<>();
     private final Set<GeoHashQuery> outstandingQueries = new HashSet<>();
+    private final HashMap<String, DataSnapshot> dataCache = new HashMap<>();
```

Comment:
It seems like this is parallel to `locationInfos`.  One is `String-->Geolocation` and the other one is `String-->DataSnapshot`.

What if we consolidated into a single `Map<String,Pair<GeoLocation,DataSnapshot>>` (or similar) so that the two maps never get out of sync?